src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs

```diff
@@ -0,0 +1,1564 @@
+                            {
+                                // The chain of entries forms a loop, which means a concurrent update has happened.
+                                ThrowHelper.ThrowInvalidOperationException_ConcurrentOperationsNotSupported();
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // Object type: Shared Generic, EqualityComparer<TValue>.Default won't devirtualize (https://github.com/dotnet/runtime/issues/10050),
+                        // so cache in a local rather than get EqualityComparer per loop iteration.
+                        EqualityComparer<T> defaultComparer = EqualityComparer<T>.Default;
+                        int i = GetBucketValue(hashCode) - 1; // Value in _buckets is 1-based
+                        while (i >= 0)
+                        {
+                            ref Entry entry = ref entries[i];
+                            if (entry.HashCode == hashCode && defaultComparer.Equals(entry.Value, item))
+                            {
+                                return i;
+                            }
+                            i = entry.Next;
+
+                            collisionCount++;
+                            if (collisionCount > (uint)entries.Length)
+                            {
+                                // The chain of entries forms a loop, which means a concurrent update has happened.
+                                ThrowHelper.ThrowInvalidOperationException_ConcurrentOperationsNotSupported();
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    int hashCode = item != null ? comparer.GetHashCode(item) : 0;
+                    int i = GetBucketValue(hashCode) - 1; // Value in _buckets is 1-based
+                    while (i >= 0)
+                    {
+                        ref Entry entry = ref entries[i];
+                        if (entry.HashCode == hashCode && comparer.Equals(entry.Value, item))
+                        {
+                            return i;
+                        }
+                        i = entry.Next;
+
+                        collisionCount++;
+                        if (collisionCount > (uint)entries.Length)
+                        {
+                            // The chain of entries forms a loop, which means a concurrent update has happened.
+                            ThrowHelper.ThrowInvalidOperationException_ConcurrentOperationsNotSupported();
+                        }
+                    }
+                }
+            }
+
+            return -1;
+        }
+
+        /// <summary>Gets the index in the bucket corresponding to the specified hashcode.</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetBucketValue(int hashCode)
```

Comment:
Why do we need this second GetBucket* method? Can't we use GetBucket[Ref] instead?

Reply:
It made a measurable difference on throughout and was visible in the asm to use GetBucketValue rather than at the usage location dereferencing the result of GetBucketRef. I didn't want both, but at least when I tested it not having both led to regressions. There may be a JIT issue, or it could have been some issue in the exact consumption pattern. I can measure it again.

Reply:
> when I tested it not having both led to regressions

Interesting observation.  That means we may be losing some perf in `Dictionary<K,V>` as well.

Reply:
I suspect we're losing perf in Dictionary for other reasons as well, but I've not yet had the time to test. If I can repro the issue I was seeing, I'll open an issue for it; otherwise I'll consolidate.